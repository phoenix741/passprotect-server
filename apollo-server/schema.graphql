scalar Date

# Encrypted content
type EncryptedContent {
  # Content encrypted with salt
  content: String!
  # Authentification tag
  authTag: String!
}

# Encrypted content
input EncryptedContentInput {
  # Content encrypted with salt
  content: String!
  # Authentification tag
  authTag: String!
}

type Error {
  fieldName: String,
  message: String
}

type Errors {
  errors: [Error]!
}

# Master key encrypted for the user
type EncryptedMasterKey {
  # Salt added to encrypt the master key
  salt: String!
  # Encrypted master key
  encryptedKey: EncryptedContent!
}

# Master key encrypted for the user
input EncryptedMasterKeyInput {
  # Salt added to encrypt the master key
  salt: String!
  # Encrypted master key
  encryptedKey: EncryptedContentInput!
}

# User defined in the application
type User {
  # Login of the user
  _id: ID!,
  # Encrypted master key
  encryption: EncryptedMasterKey!,
  # Creation date of the user
  creationAt: Date,
  # Role of the user (can be user or admin)
  role: String
  # Wallet lines of the user.
  lines: [WalletLine]!
}

input RegistrationUser {
  # Login of the user
  _id: ID!,
  # Password used for registration
  password: String!
  # Encrypted master key
  encryption: EncryptedMasterKeyInput!,
}

union RegisterUserResult = User | Errors

# Wallet line content encrypted
type EncryptedWalletLine {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContent!
}

# Wallet line content encrypted
input EncryptedWalletLineInput {
  # Salt used to encrypt wallet line
  salt: String!
  # Information of the wallet encrypted
  informations: EncryptedContentInput!
}

# Wallet
type WalletLine {
  # Id of the line
  _id: ID!
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLine!,
  # Owner of the list
  user: User!,
  # Last updated date of the line
  updatedAt: Date,
  # Revision (modification count) of the line
  _rev: Int
}

# Wallet line used as input
input WalletLineInput {
  # Id of the line, if not filled application create a new line
  _id: ID
  # Type of the line (text, password, card)
  type: String!,
  # Label of the wallet line
  label: String!,
  # Group of the wallet line
	group: String,
  # Encrypted content of the line (encrypted client side)
  encryption: EncryptedWalletLineInput!,
  # Revision (modification count) of the line, used to ensure modification of the last line
  _rev: Int
}

union CreateUpdateLineResult = WalletLine | Errors

# Information needed to connect to the server and retreive object from database.
input ConnectionInformationInput {
  # Login of the user
  username: String!,
  # Password used for connection
  password: String!
}

# Result of the connection
type ConnectionInformation {
  # Token to be used in connection information
  token: String,
  # The connected user
  user: User!
}

union CreateSessionResult = ConnectionInformation | Errors

# Wallet transaction, used to synchronize lines between devices
type WalletTransaction {
  # Id of the transaction
  _id: ID!,
  # Type of the object modified (only line)
  type: String!,
  # Line
  line: WalletLine,
  # Owner of the transaction
  user: User!,
  # The line before the modification
  before: WalletLine,
  # The line after the modification
  after: WalletLine,
  # The modification date
  updatedAt: Date!,
  # A SHA-512 calculate on the modification.
  sha512: String!
}

type RootQuery {
  # Request a user
  user(id: ID!): User,

    # Request the list of lines from the connected users.
  lines: [WalletLine],

  # Request a wallet line
  line(id: ID!): WalletLine,

	# Request all groups available on lines
	groups: [String!]

  # Request the current session
  session: User

  # Request all transactions that happen since the last synchronisation date.
  transactions(earliest: Date!): [WalletTransaction]
}

type RootMutation {
  registerUser(input: RegistrationUser!): RegisterUserResult

  createUpdateLine(input: WalletLineInput!): CreateUpdateLineResult
  removeLine(id: ID): Errors

  createSession(input: ConnectionInformationInput!): CreateSessionResult
}

type RootSubscription {
  transactionAdded: WalletTransaction!
}

schema {
  query: RootQuery
  mutation: RootMutation
  subscription: RootSubscription
}
