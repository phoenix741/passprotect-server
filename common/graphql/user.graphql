# Master key encrypted for the user
type EncryptedMasterKey {
  # Salt added to encrypt the master key
  salt: String!
  # Encrypted master key
  encryptedKey: EncryptedContent!
}

# Master key encrypted for the user
input EncryptedMasterKeyInput {
  # Salt added to encrypt the master key
  salt: String!
  # Encrypted master key
  encryptedKey: EncryptedContentInput!
}

# User defined in the application
type User {
  # Login of the user
  _id: ID!,
  # Encrypted master key
  encryption: EncryptedMasterKey!,
  # Creation date of the user
  creationAt: Date,
  # Role of the user (can be user or admin)
  role: String
  # Wallet lines of the user.
  lines: [WalletLine]!
}

input RegistrationUser {
  # Login of the user
  _id: ID!,
  # Password used for registration
  password: String!
  # Encrypted master key
  encryption: EncryptedMasterKeyInput!,
}

type RootQuery {
  # Request the list of users from the database.
  users: [User],

  # Request a user
  user(id: ID!): User,
}

union RegisterUserResult = User | Errors

type RootMutation {
  registerUser(input: RegistrationUser!): RegisterUserResult
}
