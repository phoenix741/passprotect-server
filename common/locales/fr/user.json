{
  "login": {
    "dialog": {
      "title": "Connexion",
      "button": {
        "connect": "Continuer",
        "noaccount": "Vous n'avez pas encore de compte ? ",
        "signup": "Enregistrez-vous"
      }
    },
    "form": {
      "username": {
        "field": "Votre nom d'utilisateur",
        "help": {
          "cantbeblank": "Le nom de l'utilisateur ne peut être vide.",
          "invalid": "L'utilisateur ou le mot de passe sont incorrect.",
          "unknown": "Une erreur est survenue lors de l'authentification de l'utilisateur."
        }
      },
      "password": {
        "field": "Votre mot de passe",
        "help": {
          "default": "Vous avez oublié votre mot de passe ...",
          "cantbeblank": "Le mot de passe ne peut être vide."
        }
      }
    }
  },
  "register": {
    "form": {
      "title": "Nouveau compte",
      "identity_username": {
        "field": "Nom d'utilisateur"
      },
      "identity_password1": {
        "field": "Mot de passe"
      },
      "identity_password2": {
        "field": "Mot de passe (encore)"
      },
      "validation": {
        "field": "S'enregistrer"
      }
    },
    "flash": {
      "user_created": "Votre compte a bien été enregistré"
    }
  },
  "toolbar": {
    "notlogged": "Non connecté"
  }
}
