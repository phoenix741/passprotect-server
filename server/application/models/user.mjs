import {isString} from 'lodash'
import {promise as dbPromise} from '../utils/db'
import i18n from 'i18next'

import { processMongoException, NotFoundError } from './exception'

/*
 * Model of a user contains :
 *        _id: {type: String, required: true, unique: true, lowercase: true},
 *        password: {type: String, required: true, minlength: 8},
 *        lastLogin: {type: Date},
 *        confirmationToken: {type: String, sparse: true, unique: true},
 *        passwordRequestedAt: {type: Date}
 *        encryptionKey: {type: String},
 *        encryptionIV: {type: String}
 *
 * The encryptionKey is crypted with the password of the user and a salt.
 */
export function getUsers (filter = {}) {
  const find = {}

  if (isString(filter.confirmationToken)) {
    find.confirmationToken = filter.confirmationToken
  }

  return dbPromise
    .then(db => db.collection('users').find(find).toArray())
}

export function getUser (id) {
  return dbPromise
    .then(db => db.collection('users').findOne({ _id: id.toLowerCase() }))
    .then(user => processNotFound(id, user))
}

export function registerUser (user) {
  normalizeUser(user)

  return dbPromise
    .then(db => db.collection('users').insert(user))
    .then(() => user)
    .catch(processMongoException)
}

export function saveUser (user) {
  normalizeUser(user)

  return dbPromise
    .then(db => db.collection('users').save(user))
    .then(() => user)
    .catch(processMongoException)
}

function normalizeUser (user) {
  // The _id of the user should always in lowercase to identify quickly the user
  user._id = user._id.toLowerCase()
  if (!user.role) {
    user.role = 'user'
  }
}

function processNotFound (userId, user) {
  if (!user) {
    throw new NotFoundError(i18n.t('error:user.404.userNotFound', { userId: userId }))
  }
  return user
}
