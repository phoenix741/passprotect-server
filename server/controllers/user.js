'use strict';

const config = require('config');
const _ = require('underscore');
const router = require('express-promise-router')();
const i18n = require('i18next');
const authenticate = require('../utils/passport').authenticate;
const permission = require('../utils/passport').permission;

module.exports = function (app, services) {
	/**
	 * @apiDefine ParamsUpdate
	 *
	 * @apiParam {String} _id User name
	 * @apiParam {Object} encryption Informations used for encrypted data
	 * @apiParam {String} encryption.salt Salt added to encrypt the master key
	 * @apiParam {Object} encryption.encryptedKey Encrypted master key
	 * @apiParam {String} encryption.encryptedKey.content Master key encrypted with salt and password
	 * @apiParam {String} encryption.encryptedKey.authTag Authentification tag of the master key
	 * @apiParam {String} password Password of the user
	 *
	 * @apiParamExample {json} User:
	 *   {
	 *     "_id": "myusername",
	 *     "password": "mypassword"
	 *   }
	 */

	/**
	 * @apiDefine ResponseUpdateComplete
	 *
	 * @apiSuccess {String} _id User name
	 * @apiSuccess {Object} encryption Informations used for encrypted data
	 * @apiSuccess {String} encryption.salt Salt added to encrypt the master key
	 * @apiSuccess {Object} encryption.encryptedKey Encrypted master key
	 * @apiSuccess {String} encryption.encryptedKey.content Master key encrypted with salt and password
	 * @apiSuccess {String} encryption.encryptedKey.authTag Authentification tag of the master key
	 * @apiSuccess {String} role Role of the user
	 *
	 * @apiSuccessExample {json} Success:
	 *   HTTP/1.1 200 OK
	 *   {
	 *     "_id": "myusername",
	 *     "encryption": {
	 *         "salt": "395f68144ab82169",
	 *         "encryptedKey": {
	 *             "content": "7944424037f5c534a75e70b599eb82ec7d4b02ce6064773f661a7bc1114cb1d97627ed1ad701875a8aa3a11e2b6f3e36792168a8b6327b9158907d8414bba710",
	 *             "authTag": "c2b485b1a20f025b3753e6b51e0a14e8"
	 *         }
	 *     },
	 *     "role": "user"
	 *   }
	 */
	router.route('')
	/**
	 * @api {get} /api/users Request all users

	 * @apiDescription Request the list of users from the database. Only non-sensitive informations are available. Only
	 * the adminstrator can access to this API.
	 *
	 * @apiName GetUsers
	 * @apiGroup User
	 * @apiPermission ROLE_ADMIN
	 *
	 * @apiParam {Number} [offset=0] Offset for pagination
	 * @apiParam {Number} [limit=10000] Limit of users used for pagination
	 *
	 * @apiSuccess {String} _id User name.
	 * @apiSuccess {String="user","admin"} role Role of the user.
	 *
	 * @apiSuccessExample Success:
	 *     HTTP/1.1 200 OK
	 *     [
	 *       {
	 *         "_id": "test",
	 *         "role": "user"
	 *       },
	 *       {
	 *         "_id": "test2",
	 *         "role": "user"
	 *       },
	 *     ]
	 *
	 * @apiError Unauthorized Only a connected admin can see informations
	 * @apiError Forbidden Only the admin role can access to this informations
	 *
	 * @apiErrorExample {json} Unauthorized:
	 *     HTTP/1.1 401 Unauthorized
	 *     {
	 *       "code": 401,
	 *       "message": "Unauthorized"
	 *     }
	 *
	 * @apiErrorExample {json} Forbidden:
	 *     HTTP/1.1 403 Forbidden
	 *     {
	 *       "code": 403,
	 *       "message": "Forbidden"
	 *     }
	 */
		.get(authenticate(), permission(['admin']), (req, res) => {
			const offset = req.query.offset || 0;
			const limit = req.query.limit || 10;

			return services.user.getUsers({offset, limit}).map(filterUser).then(res.json.bind(res));
		})
		/**
		 * @api {post} /api/users Register a new user
		 * @apiName PostUsers
		 * @apiGroup User
		 *
		 * @apiUse ParamsUpdate
		 * @apiUse ResponseUpdateComplete
		 */
		.post((req, res) => {
			const data = _.pick(req.body, '_id', 'password', 'encryption');

			const validationError = new Error();
			validationError.status = 400;
			if (!_.isString(data._id) || _.isEmpty(data._id)) {
				validationError.message = i18n.t('error:user.400._id');
				return Promise.reject(validationError);
			}

			if (!_.isString(data.password) || _.isEmpty(data.password) || data.password.length < 8) {
				validationError.message = i18n.t('error:user.400.password');
				return Promise.reject(validationError);
			}

			if (!_.isObject(data.encryption)) {
				validationError.message = i18n.t('error:user.400.encryption');
				return Promise.reject(validationError);
			}

			data.encryption = _.pick(data.encryption, 'salt', 'encryptedKey');

			if (!_.isObject(data.encryption.encryptedKey) || !_.isString(data.encryption.salt)) {
				validationError.message = i18n.t('error:user.400.encryption');
				return Promise.reject(validationError);
			}

			data.encryption.encryptedKey = _.pick(data.encryption.encryptedKey, 'content', 'authTag');

			if (!_.isString(data.encryption.encryptedKey.content) || !_.isString(data.encryption.encryptedKey.authTag)) {
				validationError.message = i18n.t('error:user.400.encryption');
				return Promise.reject(validationError);
			}

			return services.user.registerUser(data).then(function (user) {
				res.json(filterUser(user));
			});
		});

	/**
	 * @api {get} /api/users/:userId Request a user
	 * @apiName GetUser
	 * @apiGroup User
	 * @apiPermission ROLE_USER
	 *
	 * @apiParam {String} userId The user id
	 *
	 * @apiUse ResponseUpdateComplete
	 *
	 * @apiError Unauthorized Only a connected user can call this method
	 * @apiError Forbidden Only the user connected can access to it's information
	 *
	 * @apiErrorExample {json} Unauthorized:
	 *     HTTP/1.1 401 Unauthorized
	 *     {
	 *       "code": 401,
	 *       "message": "Unauthorized"
	 *     }
	 *
	 * @apiErrorExample {json} Forbidden:
	 *     HTTP/1.1 403 Forbidden
	 *     {
	 *       "code": 403,
	 *       "message": "Forbidden"
	 *     }
	 */
	router.route('/:userId')
		.get(authenticate(), permission(), function (req, res) {
			if (req.params.userId !== req.user._id) {
				const error = new Error('Forbidden');
				error.status = 403;
				return Promise.reject(error);
			}

			return res.json(filterUser(req.user));
		});

	app.use('/api/users', router);

	function filterUser(user) {
		return _.omit(user, 'password', 'confirmationToken');
	}
};
