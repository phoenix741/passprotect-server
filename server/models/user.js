'use strict';

const _ = require('underscore');
const db = require('../utils/db');
const i18n = require('i18next');

const processMongoException = require('./exception').processMongoException;
const NotFoundError = require('./exception').NotFoundError;

/*
 * Model of a user contains :
 *        _id: {type: String, required: true, unique: true, lowercase: true},
 *        password: {type: String, required: true, minlength: 8},
 *        lastLogin: {type: Date},
 *        confirmationToken: {type: String, sparse: true, unique: true},
 *        passwordRequestedAt: {type: Date}
 *        encryptionKey: {type: String},
 *        encryptionIV: {type: String}
 *
 * The encryptionKey is crypted with the password of the user and a salt.
 */
module.exports = function () {
	return {
		getUsers(filter, offset, limit) {
			const find = {};

			if (_.isString(filter.confirmationToken)) {
				find.confirmationToken = filter.confirmationToken;
			}

			return db.promise.then(db => {
				return Promise.fromCallback(cb => db.collection('users').find(find).skip(offset).limit(limit).toArray(cb));
			});
		},

		getUser(id) {
			return db.promise.then(db => {
				return Promise.fromCallback(cb => db.collection('users').findOne({_id: id.toLowerCase()}, cb));
			}).then(_.partial(processNotFound, id));
		},

		registerUser(user) {
			normalizeUser(user);

			return db.promise.then(db => {
				return Promise.fromCallback(cb => db.collection('users'). insert(user, cb)).return(user);
			}).catch(processMongoException);
		},

		saveUser(user) {
			normalizeUser(user);

			return db.promise.then(db => {
				return Promise.fromCallback(cb => db.collection('users').save(user, cb)).return(user);
			}).catch(processMongoException);
		}
	};

	function normalizeUser(user) {
		// The _id of the user should always in lowercase to identify quickly the user
		user._id = user._id.toLowerCase();
		if (!user.role) {
			user.role = 'user';
		}
	}

	function processNotFound(userId, user) {
		if (!user) {
			throw new NotFoundError(i18n.t('error:user.404.userNotFound', {userId: userId}));
		}
		return user;
	}
};
