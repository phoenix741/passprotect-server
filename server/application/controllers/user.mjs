import {pick, omit, isString, isEmpty, isObject} from 'lodash'
import i18n from 'i18next'
import debug from 'debug'
import fs from 'fs'
import path from 'path'
import {checkPermission} from '../utils/passport'
import {getUsers, getUser, registerUser} from '../services/user'

const log = debug('App:Controllers:User')

log('Load user type definition')
export const typeDefs = [
  fs.readFileSync(path.join(__dirname, '..', '..', '..', 'common', 'graphql', 'user.graphql'), 'utf-8')
]

export const resolvers = {
  RootQuery: {
    users (obj, args, {user}) {
      return checkPermission(user, ['admin']).then(getUsers).then(users => users.map(filterUser))
    },

    user (obj, {id}, {user}) {
      return checkPermission(user, ['admin'], id).then(() => getUser(id)).then(filterUser)
    }
  },
  WalletLine: {
    user (obj, args, {user}) {
      return checkPermission(user, ['admin'], obj.user).then(() => getUser(obj.user)).then(filterUser)
    }
  },
  WalletTransaction: {
    user (obj, {id}, {user}) {
      return checkPermission(user, ['admin'], id).then(() => getUser(obj.user)).then(filterUser)
    }
  },

  RootMutation: {
    registerUser (obj, {input}) {
      return sanitizeUser(input)
        .then(data => registerUser(data))
        .then(filterUser)
        .catch(parseErrors)
    }
  },

  RegisterUserResult: {
    __resolveType (obj) {
      if (obj.errors) {
        return 'Errors'
      }

      return 'User'
    }
  }
}

function sanitizeUser (input) {
  const data = pick(input, '_id', 'password', 'encryption')

  const validationError = new Error()
  validationError.status = 400
  if (!isString(data._id) || isEmpty(data._id)) {
    validationError.message = i18n.t('error:user.400._id')
    return Promise.reject(validationError)
  }

  if (!isString(data.password) || isEmpty(data.password) || data.password.length < 8) {
    validationError.message = i18n.t('error:user.400.password')
    return Promise.reject(validationError)
  }

  if (!isObject(data.encryption)) {
    validationError.message = i18n.t('error:user.400.encryption')
    return Promise.reject(validationError)
  }

  data.encryption = pick(data.encryption, 'salt', 'encryptedKey')

  if (!isObject(data.encryption.encryptedKey) || !isString(data.encryption.salt)) {
    validationError.message = i18n.t('error:user.400.encryption')
    return Promise.reject(validationError)
  }

  data.encryption.encryptedKey = pick(data.encryption.encryptedKey, 'content', 'authTag')

  if (!isString(data.encryption.encryptedKey.content) || !isString(data.encryption.encryptedKey.authTag)) {
    validationError.message = i18n.t('error:user.400.encryption')
    return Promise.reject(validationError)
  }

  return Promise.resolve(data)
}

function filterUser (user) {
  return omit(user, 'password', 'confirmationToken')
}

function parseErrors (err) {
  return {
    errors: [{
      fieldName: err.field || 'global',
      message: err.message
    }]
  }
}
