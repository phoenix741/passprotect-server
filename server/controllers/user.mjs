import _ from 'lodash'
import i18n from 'i18next'
import debug from 'debug'
import fs from 'fs'
import path from 'path'
import { checkPermission } from '../utils/authentification'
import { getUser, registerUser } from '../services/user'

const log = debug('App:Controllers:User')
const __dirname = path.dirname(new URL(import.meta.url).pathname);

log('Load user type definition')
export const typeDefs = [
  fs.readFileSync(path.join(__dirname, '..', 'graphql', 'user.graphql'), 'utf-8')
]

export const resolvers = {
  RootQuery: {
    async user (obj, { id }, { user }) {
      checkPermission(user, id)
      const dbUser = await getUser(id)
      return filterUser(dbUser)
    }
  },
  WalletLine: {
    async user (obj, args, { user }) {
      checkPermission(user, obj.user)
      const dbUser = await getUser(obj.user)
      return filterUser(dbUser)
    }
  },
  WalletTransaction: {
    async user (obj, { id }, { user }) {
      checkPermission(user, id)
      const dbUser = getUser(obj.user)
      return filterUser(dbUser)
    }
  },

  RootMutation: {
    async registerUser (obj, { input }) {
      try {
        const data = sanitizeUser(input)
        const user = await registerUser(data)
        return filterUser(user)
      } catch (err) {
        return parseErrors(err)
      }
    }
  },

  RegisterUserResult: {
    __resolveType (obj) {
      if (obj.errors) {
        return 'Errors'
      }

      return 'User'
    }
  }
}

function sanitizeUser (input) {
  const data = _.pick(input, '_id', 'password', 'encryption')

  const validationError = new Error()
  validationError.status = 400
  if (!_.isString(data._id) || _.isEmpty(data._id)) {
    validationError.message = i18n.t('error:user.400._id')
    throw validationError
  }

  if (!_.isString(data.password) || _.isEmpty(data.password) || data.password.length < 8) {
    validationError.message = i18n.t('error:user.400.password')
    throw validationError
  }

  if (!_.isObject(data.encryption)) {
    validationError.message = i18n.t('error:user.400.encryption')
    throw validationError
  }

  data.encryption = _.pick(data.encryption, 'salt', 'encryptedKey')

  if (!_.isObject(data.encryption.encryptedKey) || !_.isString(data.encryption.salt)) {
    validationError.message = i18n.t('error:user.400.encryption')
    throw validationError
  }

  data.encryption.encryptedKey = _.pick(data.encryption.encryptedKey, 'content', 'authTag')

  if (!_.isString(data.encryption.encryptedKey.content) || !_.isString(data.encryption.encryptedKey.authTag)) {
    validationError.message = i18n.t('error:user.400.encryption')
    throw validationError
  }

  return data
}

function filterUser (user) {
  return _.omit(user, 'password', 'confirmationToken')
}

function parseErrors (err) {
  return {
    errors: [{
      fieldName: err.field || 'username',
      message: err.message
    }]
  }
}
