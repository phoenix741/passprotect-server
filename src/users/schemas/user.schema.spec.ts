import * as mongoose from 'mongoose';
import { UserSchema, migrateV1toV2 } from './user.schema';
import { UserEntity } from '../../users/models/user.entity';

describe('LineSchema', () => {
  let UserModel: mongoose.Model<UserEntity>;

  beforeEach(() => {
    UserModel = mongoose.model('User', UserSchema);
  });

  describe('validation', () => {
    it('should not create a line - field required', async () => {
      expect.assertions(1);
      const user = new UserModel();
      try {
        await user.validate();
      } catch (err) {
        expect(Object.keys(err.errors)).toEqual(['password', '_id']);
      }
    });

    it('should create a line - all field fill', async () => {
      const user = new UserModel({
        _id: 'username',
        password: 'password',
      });

      await user.validate();
    });
  });

  describe('migrateV1toV2', () => {
    const docV1 = {
      encryption: {
        salt: 'salt',
        encryptedKey: {
          authTag: 'authTag',
          content: 'content',
        },
      },
    };

    const docV2 = {
      encryption: {
        salt: 'salt',
        iv: null,
        authTag: 'authTag',
        content: 'content',
      },
    };

    const gettersSetters = {
      get(property) {
        return this[property];
      },
      set(property, value) {
        this[property] = value;
      },
    };

    it('migrateV1toV2 - migration', async () => {
      const newDoc = Object.assign({}, gettersSetters, docV1);
      migrateV1toV2(newDoc as any as UserEntity);
      expect(newDoc).toMatchSnapshot('user migration, with migration');
    });

    it('migrateV1toV2 - no migration', async () => {
      const newDoc = Object.assign({}, gettersSetters, docV2);
      migrateV1toV2(newDoc as any as UserEntity);
      expect(newDoc).toMatchSnapshot('user migration, without migration');
    });
  });
});
