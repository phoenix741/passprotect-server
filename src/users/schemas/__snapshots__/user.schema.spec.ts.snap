// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineSchema migrateV1toV2 migrateV1toV2 - migration: user migration, with migration 1`] = `
Object {
  "encryption": Object {
    "authTag": "authTag",
    "content": "content",
    "iv": null,
    "salt": "salt",
  },
  "get": [Function],
  "set": [Function],
}
`;

exports[`LineSchema migrateV1toV2 migrateV1toV2 - no migration: user migration, without migration 1`] = `
Object {
  "encryption": Object {
    "authTag": "authTag",
    "content": "content",
    "iv": null,
    "salt": "salt",
  },
  "get": [Function],
  "set": [Function],
}
`;
